tls/s2n_aead.c

```diff
@@ -33,3 +33,19 @@ int s2n_aead_aad_init(const struct s2n_connection *conn, uint8_t * sequence_numb
 
     return 0;
 }
+
+/* Prepares an AAD for a TLS 1.3 AEAD record */
+int s2n_tls13_aead_aad_init(uint16_t record_length, uint8_t tag_length, struct s2n_stuffer *ad)
+{
+    gte_check(record_length, 0);
+    gt_check(tag_length, 0);
+    notnull_check(ad);
+
+    /* tls1.3 ad = content type || version || length */
+    GUARD(s2n_stuffer_write_uint8(ad, TLS_APPLICATION_DATA)); /* fixed to 0x17 */
+    GUARD(s2n_stuffer_write_uint8(ad, 3)); /* TLS record layer              */
+    GUARD(s2n_stuffer_write_uint8(ad, 3)); /* version fixed at 1.2 (0x0303) */
+    GUARD(s2n_stuffer_write_uint16(ad, record_length + tag_length));
```

Comment:
legacy_record_version:  The legacy_record_version field is always
      0x0303.  TLS 1.3 TLSCiphertexts are not generated until after
      TLS 1.3 has been negotiated, so there are no historical
      compatibility concerns where other values might be received.  Note
      that the handshake protocol, including the ClientHello and
      ServerHello messages, authenticates the protocol version, so this
      value is redundant.

   length:  The length (in bytes) of the following
      TLSCiphertext.encrypted_record, which is the sum of the lengths of
      the content and the padding, plus one for the inner content type,
      plus any expansion added by the AEAD algorithm.  The length
      MUST NOT exceed 2^14 + 256 bytes.  An endpoint that receives a
      record that exceeds this length MUST terminate the connection with
      a "record_overflow" alert.

Reply:
actually checking it in the aad construction effectively checks it for both construction and parsing, since an aad needs to match up with the record ciphertext :)